File: src/services/agent-registry.ts
import type { Agent } from "../agent.js";
import type { MessageStore } from "../messages.js";

export type AgentStatus = "idle" | "running" | "completed" | "error";

export interface AgentSession {
  id: string;
  type: "main" | "sub";
  agent: Agent;
  store: MessageStore;
  status: AgentStatus;
  task?: string;
  parentId?: string;
  summary?: string;
  tokenCount?: number;
  toolCalls?: number;
}

export type AgentRegistryUpdateCallback = (sessions: AgentSession[]) => void;

export class AgentRegistry {
  private sessions: Map<string, AgentSession> = new Map();
  private nextSubId = 2;
  private updateCallback?: AgentRegistryUpdateCallback;

  setUpdateCallback(callback: AgentRegistryUpdateCallback): void {
    this.updateCallback = callback;
  }

  private notifyUpdate(): void {
    this.updateCallback?.(this.getAll());
  }

  register(session: AgentSession): void {
    this.sessions.set(session.id, session);
    this.notifyUpdate();
  }

  get(id: string): AgentSession | undefined {
    return this.sessions.get(id);
  }

  getAll(): AgentSession[] {
    return Array.from(this.sessions.values()).sort(
      (a, b) => parseInt(a.id, 10) - parseInt(b.id, 10),
    );
  }

  updateStatus(id: string, status: AgentStatus): void {
    const session = this.sessions.get(id);
    if (session) {
      session.status = status;
      this.notifyUpdate();
    }
  }

  updateSummary(id: string, summary: string): void {
    const session = this.sessions.get(id);
    if (session) {
      session.summary = summary;
      this.notifyUpdate();
    }
  }

  updateProgress(
    id: string,
    updates: { tokenCount?: number; toolCalls?: number },
  ): void {
    const session = this.sessions.get(id);
    if (session) {
      if (updates.tokenCount !== undefined)
        session.tokenCount = updates.tokenCount;
      if (updates.toolCalls !== undefined)
        session.toolCalls = updates.toolCalls;
      this.notifyUpdate();
    }
  }

  allocateSubId(): string {
    for (let i = 2; i <= 9; i++) {
      const id = String(i);
      if (!this.sessions.has(id)) {
        this.nextSubId = i + 1 > 9 ? 2 : i + 1;
        return id;
      }
    }
    for (let i = 2; i <= 9; i++) {
      const id = String(i);
      const session = this.sessions.get(id);
      if (
        session &&
        session.type === "sub" &&
        (session.status === "completed" || session.status === "error")
      ) {
        this.sessions.delete(id);
        this.nextSubId = i + 1 > 9 ? 2 : i + 1;
        this.notifyUpdate();
        return id;
      }
    }
    const id = String(this.nextSubId);
    this.nextSubId++;
    if (this.nextSubId > 9) this.nextSubId = 2;
    return id;
  }

  remove(id: string): void {
    this.sessions.delete(id);
    this.notifyUpdate();
  }

  clear(): void {
    this.sessions.clear();
    this.nextSubId = 2;
  }
}

File: internal/agent_registry.go
package internal

import (
	"fmt"
	"sync"
)

// AgentSessionType defines the role of an agent.
type AgentSessionType string

const (
	AgentMain AgentSessionType = "main"
	AgentSub  AgentSessionType = "sub"
)

// AgentStatus defines the execution state of an agent.
type AgentStatus string

const (
	StatusIdle      AgentStatus = "idle"
	StatusRunning   AgentStatus = "running"
	StatusCompleted AgentStatus = "completed"
	StatusError     AgentStatus = "error"
)

// AgentSession tracks an agent's lifecycle and metadata.
type AgentSession struct {
	ID       string
	Type     AgentSessionType
	Agent    *Agent
	Status   AgentStatus
	Task     string
	ParentID string
	Summary  string
}

// AgentRegistry manages multiple agent sessions.
type AgentRegistry struct {
	mu             sync.Mutex
	sessions       map[string]*AgentSession
	order          []string
	activeID       string
	updateCallback func(sessions []AgentSession)
}

// NewAgentRegistry creates an agent registry with the initial agent.
func NewAgentRegistry(primary *Agent) *AgentRegistry {
	s := &AgentSession{
		ID:     "1",
		Type:   AgentMain,
		Agent:  primary,
		Status: StatusIdle,
	}
	return &AgentRegistry{
		sessions: map[string]*AgentSession{"1": s},
		order:    []string{"1"},
		activeID: "1",
	}
}

// Register adds a new session and notifies.
func (r *AgentRegistry) Register(s *AgentSession) {
	r.mu.Lock()
	r.sessions[s.ID] = s
	r.order = append(r.order, s.ID)
	r.mu.Unlock()
	r.notify()
}

// Get returns a session by ID.
func (r *AgentRegistry) Get(id string) (*AgentSession, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	s, ok := r.sessions[id]
	return s, ok
}

// List returns all registered sessions.
func (r *AgentRegistry) List() []AgentSession {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := make([]AgentSession, 0, len(r.order))
	for _, id := range r.order {
		if s, ok := r.sessions[id]; ok {
			out = append(out, *s)
		}
	}
	return out
}

// Active returns the currently active agent.
func (r *AgentRegistry) Active() *Agent {
	r.mu.Lock()
	defer r.mu.Unlock()
	if s, ok := r.sessions[r.activeID]; ok {
		return s.Agent
	}
	return nil
}

// NextActive cycles to the next agent and returns it.
func (r *AgentRegistry) NextActive() *Agent {
	r.mu.Lock()
	defer r.mu.Unlock()
	if len(r.order) <= 1 {
		return nil
	}
	idx := -1
	for i, id := range r.order {
		if id == r.activeID {
			idx = i
			break
		}
	}
	nextIdx := (idx + 1) % len(r.order)
	r.activeID = r.order[nextIdx]
	return r.sessions[r.activeID].Agent
}

// AllocateSubID returns a unique ID for a sub-agent.
func (r *AgentRegistry) AllocateSubID() string {
	r.mu.Lock()
	defer r.mu.Unlock()
	for i := 2; i <= 9; i++ {
		id := fmt.Sprintf("%d", i)
		if _, ok := r.sessions[id]; !ok {
			return id
		}
	}
	return fmt.Sprintf("%d", len(r.sessions)+1)
}

// SetUpdateCallback sets a function called whenever sessions change.
func (r *AgentRegistry) OnUpdate(cb func(sessions []AgentSession)) {
	r.mu.Lock()
	r.updateCallback = cb
	r.mu.Unlock()
}

func (r *AgentRegistry) notify() {
	r.mu.Lock()
	cb := r.updateCallback
	r.mu.Unlock()
	if cb != nil {
		cb(r.List())
	}
}

// UpdateStatus sets the status of a session.
func (r *AgentRegistry) UpdateStatus(id string, status AgentStatus, summary string) {
	r.mu.Lock()
	if s, ok := r.sessions[id]; ok {
		s.Status = status
		if summary != "" {
			s.Summary = summary
		}
	}
	r.mu.Unlock()
	r.notify()
}
